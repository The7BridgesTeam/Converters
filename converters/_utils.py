import sys
from typing import Callable, Union

from pydash import merge


# --------------------------------------------------------------------------------------------------
# Decorators
# --------------------------------------------------------------------------------------------------
class class_or_instance_method:  # noqa
    """This decorator allows you to make a method act on a class or an
    instance. So:

    >>> class MyClass(object):
    ...     @class_or_instance_method
    ...     def add_property(cls_or_self, prop, val):
    ...         setattr(cls_or_self, prop, val)
    >>> MyClass.a
    Traceback (most recent call last):
    ...
    AttributeError: type object 'MyClass' has no attribute 'a'
    >>> MyClass.add_property('a', 1)
    >>> MyClass.a
    1
    >>> inst = MyClass()
    >>> inst.a
    1
    >>> inst.b
    Traceback (most recent call last):
    ...
    AttributeError: 'MyClass' object has no attribute 'b'
    >>> inst.add_property('b', 2)
    >>> inst.b
    2
    >>> MyClass.b
    Traceback (most recent call last):
    ...
    AttributeError: type object 'MyClass' has no attribute 'b'
    """

    def __init__(self, func):
        self._func = func

    def __get__(self, obj, cls):
        target = obj if obj is not None else cls

        def class_or_instance_wrapper(*args, **kwargs):
            return self._func(target, *args, **kwargs)

        return class_or_instance_wrapper


# --------------------------------------------------------------------------------------------------
# Descriptors
# --------------------------------------------------------------------------------------------------
class MergedClassProperty:
    def __init__(self, class_property):
        self.class_property = class_property
        self.class_cache_property = f'__{class_property}_cache'

    def __get__(self, instance, cls):
        # need to get it direct in case a superclass has it set differently
        cached_result = cls.__dict__.get(self.class_cache_property)
        if cached_result is not None:
            return cached_result

        merged_prop = merge(
            {},
            *(getattr(super_class, self.class_property, {}) for super_class in cls.__mro__[-1::-1]),
        )
        setattr(cls, self.class_cache_property, merged_prop)
        return merged_prop


# --------------------------------------------------------------------------------------------------
# Functions
# --------------------------------------------------------------------------------------------------
def cn(obj):
    if isinstance(obj, type):
        return obj.__name__
    return type(obj).__name__


def identity(arg):
    return arg


def infinite_counter(initial=1, increment=1):
    counter = initial
    while True:
        yield counter
        counter += increment


def yield_forever(value):
    while True:
        yield value


def is_union_type(tp):
    if sys.version_info[:3] >= (3, 7, 0):  # PEP 560
        from typing import _GenericAlias
        return tp is Union or isinstance(tp, _GenericAlias) and tp.__origin__ is Union

    __origin__ = getattr(tp, '__origin__', False)
    return tp is typing.Union or __origin__ and __origin__ is Union


def is_one_of_union_types(source, union: Union):
    return any(type(source) is ut or isinstance(source, ut) for ut in union.__args__)
