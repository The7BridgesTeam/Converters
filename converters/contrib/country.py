import logging
import re

from django.conf import settings

from django_countries import countries
from django_countries.fields import Country as DJCountry

from ..converter import Converter


log = logging.getLogger(__name__)


def _country_name_normalizer(country_name):
    return re.sub(r'\s+', ' ', country_name.lower().strip())


def _populate_country_mapping():
    country_name_to_code = {_country_name_normalizer(name): code for code, name in countries}

    country_name_to_code.update(
        {
            _country_name_normalizer(alpha3): code
            for code, (alpha3, _) in countries.alt_codes.items()
        }
    )

    for code, alias_names in getattr(settings, 'COUNTRY_ALIASES', {}).items():
        for alias in alias_names:
            country_name_to_code[_country_name_normalizer(alias)] = code

    return country_name_to_code


COUNTRY_NORMALIZED_NAME_TO_CODE_MAP = _populate_country_mapping()


def convert_to_country_code(country_name):
    if country_name is None:
        return ''

    if isinstance(country_name, DJCountry):
        return country_name.code

    if not isinstance(country_name, str):
        raise TypeError(f'Invalid type for country name: {type(country_name)}')

    normalized_name = _country_name_normalizer(country_name)

    if not normalized_name:
        return ''

    if normalized_name.upper() in COUNTRY_NORMALIZED_NAME_TO_CODE_MAP.values():
        # It's a country code already
        return normalized_name.upper()

    code_from_name = COUNTRY_NORMALIZED_NAME_TO_CODE_MAP.get(normalized_name)
    if code_from_name is not None:
        return code_from_name.upper()

    log.warning('Could not determine country code of %r', country_name)

    return ''


Converter.register_converter_for_name(convert_to_country_code, 'country_code')
