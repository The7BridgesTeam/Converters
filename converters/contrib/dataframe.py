from io import StringIO, TextIOBase

import pandas as pd

from ..converter import Converter, create_compound_converter


def is_usable_as_dtype(arg):
    if arg is None:
        return False

    try:
        pd.api.types.pandas_dtype(arg)
        return True
    except TypeError:
        return False


def dataframe_getter(inst, attr, val, opts):
    groupby = opts.get('groupby')
    if groupby is not None:
        # when groupby is specified we always pass self through, so ignore attr
        if isinstance(groupby, dict):
            return inst.groupby(**groupby)
        return inst.groupby(groupby)

    if attr == 'self':
        return inst

    if attr not in inst:
        return None

    if opts.get('multiple_columns'):
        vals = inst.filter(regex=rf'{attr}(\.[0-9]+)?$')
        if opts.get('pass_through_series', False):
            return vals.iloc[0]
        return list(vals.iloc[0])

    vals = inst[attr]
    if opts.get('pass_through_series', False):
        return vals

    if vals.unique().shape[0] > 1:
        raise ValueError('Unable to determine value from column, multiple values exist')
    return vals.iloc[0]


class DataFrameToDictConverter(Converter):
    """Convert
    pandas dataframes to dictionaries.

    The easiest way to use it is with some CSV:

        >>> csv_str = 'col1,col2,col3\\n1, 2, 3\\n4,5,6\\n1, 3, 5'
        >>> class MyRowConverter(DataFrameToDictConverter):
        ...     conversions = [('col1', int), ('col3', int)]
        >>> MyRowConverter.convert_csv(csv_str)
        [{'col1': 1, 'col3': 3}, {'col1': 4, 'col3': 6}, {'col1': 1, 'col3': 5}]

    Note that you need to use converters and/or dtype options if you don't just want strings - data
    from CSV is assumed to be string data by default to avoid pandas automatic type detection which
    can produce inconsistent results.

    But cooler things can be achieved by nesting and using groupby:

        >>> csv_str = 'col1,col2,col3\\n1, 1, 1\\n1, 1, 2\\n1, 2, 1'
        >>> class MyGroupByCol2Converter(DataFrameToDictConverter):
        ...     conversions = [
        ...         ('col2', int),
        ...         ('col3', {
        ...             'groupby': {'level': 0},
        ...             'converter': lambda x: x['col3'].values[0],
        ...             'dtype': int,
        ...         }),
        ...     ]
        >>> class MyCSVConverter(DataFrameToDictConverter):
        ...     conversions = [
        ...         ('col1', int),
        ...         ('col2s', {
        ...             'groupby': 'col2', 'converter': MyGroupByCol2Converter,
        ...         }),
        ...     ]
        >>> MyCSVConverter.convert_csv(csv_str, groupby='col1')
        [{'col1': 1, 'col2s': [{'col2': 1, 'col3': [1, 2]}, {'col2': 2, 'col3': [1]}]}]

    Note that we used `'dtype': int` explicitly for `'col3'` because `DataFrameToDictConverter`
    would not be able to deduce a dtype automatically from the converter `MyGroupByCol2Converter`.

    ## Extra `conversions` options

    ### `'dtype'`

    Use this to cause us to specify what dtype we are expecting for the column. Before passing
    the dataframe through to the standard converter routines, we call `.astype(dtype=self.dtypes)`
    on it to ensure that dtypes specified here and any that can be deduced from the converters
    are what the column actually uses.

    ## `converter_options`

    ### `'extra_dtypes'`

    If you need to reference columns in the dataframe that are not listed in `conversions`,
    (for example in your converters) and they need to be something other than a string, then use
    this to specify their types.
    """

    from_class = pd.DataFrame
    to_class = dict

    converter_options = {
        # see notes in docstring
        'extra_dtypes': {},
    }

    # ----------------------------------------------------------------------------------------------
    # Convenience methods
    # ----------------------------------------------------------------------------------------------
    @classmethod
    def convert_csv(cls, csv, **kwargs):
        if isinstance(csv, str):
            csv_io = StringIO(csv)
        elif isinstance(csv, TextIOBase):
            csv_io = csv
        else:
            raise TypeError('Unknown CSV type')

        # pandas "helpfully" automatically detects data types in columns and so the type of a column
        # may vary depending on the data in the column. This is actually pretty unhelpful, as we
        # want to be sure of the types downstream. Therefore assume all are strings, and in
        # `instance_convert` we will apply dtypes dynamically.
        return cls.convert_dataframe_of_rows(pd.read_csv(csv_io, dtype=str), **kwargs)

    @classmethod
    def convert_dataframe_of_rows(cls, df, groupby=None) -> list:
        if isinstance(groupby, str):
            groupby = {'by': groupby}
        if groupby is None:
            groupby = {'level': 0}
        return [cls.convert(row[1]) for row in df.groupby(**groupby)]

    # ----------------------------------------------------------------------------------------------
    # Subclass methods
    # ----------------------------------------------------------------------------------------------
    def instance_convert(self):
        dtypes = self.dtypes
        if len(dtypes) > 0:
            self.source = self.source.astype(dtypes)

        return super().instance_convert()

    def dynamic_copy_attrs(self):
        attrs = list(self.normalize_copy_attrs(super().dynamic_copy_attrs()))
        for (t, s, opts) in filter(lambda x: x[2].get('groupby'), attrs):
            opts['converter'] = create_compound_converter(
                lambda gb_val_df_tuple: gb_val_df_tuple[1], opts['converter'],
            )
        return attrs

    @property
    def from_getter(self):
        return dataframe_getter

    def val_is_collection(self, val):
        return not isinstance(val, (pd.DataFrame, pd.Series)) and super().val_is_collection(val)

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # Replace nans with None
        self.source = self.source.where((pd.notnull(self.source)), None)

    # ----------------------------------------------------------------------------------------------
    # Internals
    # ----------------------------------------------------------------------------------------------
    @property
    def dtypes(self):
        if hasattr(self, '_dtypes_cache'):
            return self._dtypes_cache

        self._dtypes_cache = dict(self._resolved_converter_options['extra_dtypes'])
        # Deduce dtype info from the converters as much as possible, or use the explicitly
        # configured dtype setting in the converter copy attribute options
        self._dtypes_cache.update(
            {
                from_attr: dtype
                for from_attr, dtype in (
                    (from_attr, opts.get('dtype', opts.get('converter')))
                    for (to_attr, from_attr, opts) in self.dynamic_copy_attrs()
                )
                if from_attr in self.source and is_usable_as_dtype(dtype)
            }
        )
        return self._dtypes_cache
