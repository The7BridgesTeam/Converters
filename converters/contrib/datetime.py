from dateutil import parser
from pytz import timezone

from ..converter import Converter

__all__ = [
    'convert_str_to_datetime',
    'convert_str_to_naive_datetime',
    'convert_naive_str_to_utc_datetime',
    'convert_str_to_time',
    'exports',
]


def convert_str_to_datetime(str_):
    return parser.parse(str_)


def convert_str_to_naive_datetime(str_):
    return parser.parse(str_, ignoretz=True)


def convert_naive_str_to_utc_datetime(str_):
    naive_dt = convert_str_to_naive_datetime(str_)
    return timezone('utc').localize(naive_dt)


def convert_str_to_time(str_):
    return convert_str_to_datetime(str_).time()


for func, name in [
    (convert_str_to_datetime, 'datetime'),
    (convert_str_to_time, 'time'),
    (convert_str_to_naive_datetime, 'datetime_naive'),
    (convert_naive_str_to_utc_datetime, 'datetime_naive_to_utc'),
]:
    Converter.register_converter_for_name(func, name)

exports = __all__
