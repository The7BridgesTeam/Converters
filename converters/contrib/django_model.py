import logging
from typing import List, Optional, Union

from django.core.files.base import ContentFile
from django.db.models import FileField, ManyToManyField, Model
from django.db.models.fields.files import FieldFile
from django.db.models.fields.related import ManyToManyRel, ManyToOneRel, OneToOneRel
from django.db.models.fields.reverse_related import ForeignObjectRel
from django.db.transaction import atomic

from ..converter import _SENTRY, DEDUPE_ALL, NOS, Converter


__all__ = ['DjangoModelConverter']
log = logging.getLogger(__name__)


class DjangoModelConverter(Converter):
    """Convert
    to and from django models.

    The main things this provides are:

    - Correct handling of model relationships
    - Automatic saving of source data to `FileField`s.
    - Detection of fields to copy based on the models passed in.
    - On by default saving of the instance on creating / modifying it

    More details below.

    ### `FileField` support

    If you have a model

        class MyModel(Model):
            a_file = FileField()

    And you create a converter:

        class MyModelConverter(DjangoModelConverter):
            from_class = dict
            to_class = MyModel

    By default this will save the contents of the source `a_file` attribute
    directly to file. So:

        inst = MyModelConverter.convert({'a_file': 'this text will go into the file'})
        inst.a_file.read() == b'this text will go into the file'

    If setting the file contents through a string or bytes object, as in the
    example above, the default file extension will be the field name. If you
    want to change that, use the `file_extension` option to set a specific
    value:

        class MyModelConverter(DjangoModelConverter):
            from_class = dict
            to_class = MyModel

            converter_copy_attrs = [
                ('a_file', {'file_extension': 'txt'}),
            ]

        inst = MyModelConverter.convert({'a_file': 'this text will go into the file'})
        assert inst.a_file.url.endswith('.txt')

    To remove it altogether, just use an empty string: `{'file_extension': ''}`.

    ### Automatic field copying.

    This behaviour is off by default, but you can add
    `converter_options = {'django_model_add_fields_to_copy_attrs': True}`
    to enable it, or (recommeded) inherit from `AutoDjangoModelConverter` instead.

    The fields that are automatially added will be from the `to_class`, if it
    is a model, else from the `from_class`. One of the `to_class` or `from_class`
    needs to be a model (subclass of `django.db.models.Model`).

    Creating to-many relationships from to-many relationships is not automatically
    done, because often that's not what is wanted, and it can break existing
    relationships. If it is, then you can specify those fields in
    `converter_copy_attrs` to enable them.

    So for example:

        class ChildModel(Model):
            child_text = TextField()
            parent = ForeignKey('ParentModel', related_name='children')

        class ParentModel(Model):
            parent_text = TextField()

        class ShallowParentCopier(AutoDjangoModelConverter):
            from_class = ParentModel
            to_class = ParentModel

    Since `ShallowParentCopier` does not specify recursing to the child model,
    if you do:

        parent = ParentModel.objects.create(parent_text='Parent text')
        ChildModel.objects.create(parent=parent, child_text='Child 1 text')
        ChildModel.objects.create(parent=parent, child_text='Child 2 text')
        ShallowParentCopier.convert(parent)

    You just end up with a single row in the `ParentModel` table. However, you
    can assign over the children if you specify the `'children'` attribute
    to recurse through:

        class DeepParentConverter(AutoDjangoModelConverter):
            from_class = ParentModel
            to_class = ParentModel
            converter_copy_attrs = [
                'children'
            ]

        DeepParentConverter.convert(parent)

    And you will end up with a new `ParentModel` and the two existing children
    will be re-assigned to the new parent. To copy them, use a sub-converter:

        class ChildConverter(AutoDjangoModelConverter):
            from_class = ChildModel
            to_class = ChildModel

        class DeepParentConverter(AutoDjangoModelConverter):
            from_class = dict
            to_class = ParentModel
            converter_copy_attrs = [
                ('children', ChildConverter),
            ]

        DeepParentConverter.convert(parent)

    This is tested in `test_converter_django_model.py`.
    """

    converter_options = {
        'pass_attrs_to_init': True,
        'django_model_atomic': False,
        'django_model_converter_save': True,
        'django_model_add_fields_to_copy_attrs': False,
    }

    @property
    def should_save(self):
        return (
            issubclass(self.to_class, Model)
            and self._resolved_converter_options['django_model_converter_save']
        )

    @property
    def should_add_fields_to_copy_attrs(self):
        return self._resolved_converter_options['django_model_add_fields_to_copy_attrs']

    @property
    def is_to_direction(self):
        # This controls how we deduce which fields to copy.
        #
        # If we are copying to a Model we use the target's fields.
        #
        # If we are not copying to a Model but are copying from a Model
        # we should use the source's fields.
        #
        # This can be overridden by subclasses who may wish to convert Model to
        # Model and use the source Model's fields.
        return issubclass(self.to_class, Model)

    def dynamic_copy_attrs(self):
        attrs = list(self.normalize_copy_attrs(super().dynamic_copy_attrs()))

        to_is_model = issubclass(self.to_class, Model)
        from_is_model = issubclass(self.from_class, Model)

        to_fields = {} if not to_is_model else _dict_of_model_fields(self.to_class)
        from_fields = {} if not from_is_model else _dict_of_model_fields(self.from_class)

        if self.is_to_direction:
            copy_fields = to_fields
            fname_to_copy_attr_ind_map = {
                to_attr: ind for ind, (to_attr, from_attr, opts) in enumerate(attrs)
            }
        else:
            copy_fields = from_fields
            fname_to_copy_attr_ind_map = {
                from_attr: ind
                for ind, (to_attr, from_attr, opts) in enumerate(attrs)
                if from_attr != NOS
            }

        for fname in copy_fields.keys():
            if fname in {'id', 'pk'}:
                # do not copy the id or pk by default, generally we are trying
                # to make a copy of the row, not update an existing one.
                continue

            to_attrs_set = {to_attr for (to_attr, _, _) in attrs}

            # ! this mutates attrs
            self._ensure_field_is_in_copy_attrs(
                fname, fname_to_copy_attr_ind_map, attrs, to_is_model, from_is_model, to_attrs_set
            )

        return attrs

    def instantiate(self, dct):
        inst = super().instantiate(dct)
        if inst is None or inst is _SENTRY:
            return inst
        # by default save immediately after instantiating so that reverse
        # relation operations will work
        if self.should_save:
            inst.save(force_insert=True)
        return inst

    def populate_instance(self, inst, params):
        if len(params) == 0:
            # nothing to do
            return inst
        inst = super().populate_instance(inst, params)
        if self.should_save:
            inst.save()
        return inst

    def instance_convert(self):
        if self._resolved_converter_options['django_model_atomic']:
            with atomic():
                return super().instance_convert()
        return super().instance_convert()

    # ----------------------------------------------------------------------------------------------
    # Private methods
    # ----------------------------------------------------------------------------------------------
    def _ensure_field_is_in_copy_attrs(
        self, fname, fname_to_copy_attr_ind_map, attrs, to_is_model, from_is_model, to_attrs_set
    ):
        # build the options for this field, using any current ones
        if fname not in fname_to_copy_attr_ind_map:
            if not self.should_add_fields_to_copy_attrs:
                # we've been told not to add fields that aren't there.
                return

            to_name = fname
            if to_name in to_attrs_set:
                # this attribute is being copied by a different instruction, so
                # we don't need to
                return

            # if this is a to-many don't add it automatically by default, it's
            # likely not what the caller wants, and if it is they can add it
            # explicitly.
            if from_is_model:
                from_field = _dict_of_model_fields(self.from_class).get(fname)
                if _field_is_to_many(from_field):
                    return

            if to_is_model:
                to_field = _dict_of_model_fields(self.to_class).get(to_name)
                if _field_is_to_many(to_field):
                    return

            opts = {}
            from_name = fname
            attrs.append((to_name, from_name, opts))
        else:
            to_name, from_name, opts = attrs[fname_to_copy_attr_ind_map[fname]]

        if to_is_model:
            # We need to handle one to many relations by adding them
            # only after we've initialized (so that our row exists when
            # we try to set the back-refrencing foreign key on the relation),
            # and we need to use the correct setter.
            to_field = _dict_of_model_fields(self.to_class).get(to_name)
            if isinstance(to_field, ManyToOneRel):
                opts.setdefault('initialization_attribute', False)
                if not isinstance(to_field, OneToOneRel):
                    opts.setdefault('setter', _set_many_to_one_field(to_field.get_accessor_name()))
                opts.setdefault('reverse_attr_name', to_field.remote_field.name)

            if isinstance(to_field, (ManyToManyField, ManyToManyRel)):
                opts.setdefault('setter', _set_many_to_one_field(to_name))
                opts.setdefault('initialization_attribute', False)

            if isinstance(to_field, FileField):
                opts.setdefault('setter', _set_file_field)
                opts.setdefault('initialization_attribute', False)

        if from_is_model:
            # Need to convert from one to many relation manager to an actual
            # iterable of model instances we can convert with.
            from_field = _dict_of_model_fields(self.from_class).get(from_name)
            if isinstance(from_field, ManyToOneRel):

                def many_to_one_getter(src, from_attr, opts):
                    return getattr(src, from_field.get_accessor_name()).all()

                opts.setdefault('getter', many_to_one_getter)


class AutoDjangoModelConverter(DjangoModelConverter):
    converter_options = {
        'django_model_add_fields_to_copy_attrs': True,
    }

# --------------------------------------------------------------------------------------------------
# Private helper functions
# --------------------------------------------------------------------------------------------------
def _set_file_field(inst, attr, val, opts):
    if isinstance(val, FieldFile):
        # can just set directly
        setattr(inst, attr, val)
        return

    if val is None:
        setattr(inst, attr, None)
        return

    if isinstance(val, (bytes, str)):
        extension = opts.get('file_extension', attr)

        if extension and not extension.startswith('.'):
            extension = f'.{extension}'

        getattr(inst, attr).save(f'{inst.id}{extension}', ContentFile(val))
        return

    raise NotImplementedError('DjangoModelConverter can only convert str or bytes to FileField')


def _set_many_to_one_field(attr_name):
    def _set_many_to_one_field_inner(inst, attr, val):
        getattr(inst, attr_name).set(val)

    return _set_many_to_one_field_inner


def _dict_of_model_fields(model):
    return {
        (f.get_accessor_name() if isinstance(f, ForeignObjectRel) else f.name): f
        for f in model._meta.get_fields()
    }


def _field_is_to_many(field):
    return isinstance(field, ManyToManyField) or isinstance(field, ForeignObjectRel)
