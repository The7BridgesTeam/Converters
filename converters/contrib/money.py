from django.conf import settings

from moneyed import Money
from pydash import get

from ..converter import _SENTRY


def currency_converter_for_field(field, fallback_currency_field=None):
    def currency_converter(src):
        amount = get(src, field, _SENTRY)
        if isinstance(amount, Money):
            return amount

        currency = get(src, f'{field}_currency')
        if amount is None or amount is _SENTRY:
            # no amount passed
            return amount

        if currency is None:
            currency = (
                get(src, fallback_currency_field) or settings.HOME_CURRENCY
                if fallback_currency_field
                else settings.HOME_CURRENCY
            )

        return Money(amount, currency)

    return currency_converter
