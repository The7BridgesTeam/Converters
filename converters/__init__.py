from .converter import Converter, DictConverter, InvalidToClassException, ValueRequired, NOS

__all__ = ['Converter', 'DictConverter', 'InvalidToClassException', 'NOS', 'ValueRequired']

try:
    import dateutil
    import pytz
except ImportError:
    pass
else:
    from .contrib.datetime import *
    __all__ += exports
