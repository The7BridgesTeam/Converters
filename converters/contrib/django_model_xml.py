from unidecode import unidecode

from .django_model import DjangoModelConverter
from .xml import XMLConverter


class DjangoModelXMLConverter(DjangoModelConverter, XMLConverter):
    converter_options = {'convert_nones_to_blanks': True}


class ASCIIDjangoModelXMLConverter(DjangoModelXMLConverter):
    converter_washers = {str: unidecode}
