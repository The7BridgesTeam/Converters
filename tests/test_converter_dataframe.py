from io import StringIO

import pandas as pd
import pytest

from converters import NOS, Converter
from converters.contrib.dataframe import DataFrameToDictConverter
from ._test_utils import assert_deep_pattern_match


class ChildConverter(DataFrameToDictConverter):
    conversions = [
        'child_value',
    ]


@pytest.fixture
def converter_to_dict():
    class ConverterToDict(DataFrameToDictConverter):
        conversions = [
            'id',
            'parent_value',
            ('children', {'groupby': 'child_id', 'converter': ChildConverter}),
        ]

    Converter.register_converter_for_name(ConverterToDict, 'ConverterToDict')
    try:
        yield ConverterToDict
    finally:
        Converter.unregister_converter('ConverterToDict')


def test_dataframe_to_dict_converter(converter_to_dict):
    csv_data = """id,parent_value,child_id,child_value
    1,pval,1,cval
    1,pval,2,cval2
    1,pval,3,"""
    df = pd.read_csv(StringIO(csv_data))

    # Add a test that missing columns get defaulted
    converter_to_dict.conversions.append(('missing_column', NOS, 'missing column value'))

    converted = converter_to_dict.convert(df)
    assert_deep_pattern_match(
        converted,
        {
            'id': 1,
            'missing_column': 'missing column value',
            'parent_value': 'pval',
            'children': [
                {'child_value': 'cval',},
                {'child_value': 'cval2',},
                {'child_value': None,},
            ],
        },
    )


def test_errors_out_if_duplicate_values_with_no_groupby(converter_to_dict):
    # the parent_value is duplicated, which requires a 'groupby' key
    # (the same as a CSV dump from a DB query: we'll get multiple rows for one -> many
    # relations but the parent values should be consistent)
    csv_data = """id,parent_value,child_id,child_value
    1,pval,1,cval
    1,pval_error,2,cval2"""
    df = pd.read_csv(StringIO(csv_data))
    with pytest.raises(ValueError):
        converter_to_dict.convert(df)


def test_csv_multiple_columns(snapshot):
    class MyConverter(DataFrameToDictConverter):
        conversions = [
            ('cola', {'multiple_columns': True}),
            ('colb',),
        ]

    result = MyConverter.convert_csv(
        """\
cola,colb,cola,colb
a11,b11,a12,b12
a21,b21,a22,b22
"""
    )
    assert snapshot == result
    # and this is the main point, that cola declared as multiple_columns: True collects the entries
    # from each row into a list.
    assert result[0]['cola'] == ['a11', 'a12']
