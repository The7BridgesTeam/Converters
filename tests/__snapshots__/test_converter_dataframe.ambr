# name: test_csv_multiple_columns
  <class 'list'> [
    <class 'dict'> {
      'cola': <class 'list'> [
        'a11',
        'a12',
      ],
      'colb': 'b11',
    },
    <class 'dict'> {
      'cola': <class 'list'> [
        'a21',
        'a22',
      ],
      'colb': 'b21',
    },
  ]
---
